### PR TITLE
maybe: implement Apply, Applicative, Chain, and Monad

### PR DESCRIPTION
**Pull request by davidchambers**
Reading fantasyland/fantasy-land#85 improved my understanding of these types, and motivated me to add the methods necessary for Maybe to implement the Apply, Applicative, Chain, and Monad specifications.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -46,6 +46,13 @@
     throw new Error('Cannot instantiate Maybe');
   }
 
+  //  Maybe.of :: a -> m a
+  Maybe.of = function(x) {
+    return new Just(x);
```
Comment (SimonRichardson) [inline index.js]: :+1:

File: index.js [hunk 2]
```diff
@@ -55,6 +62,11 @@
   }
   extend(Nothing, Maybe);
```
Comment (buzzdecafe) [inline index.js]: maybe this is just me, but shouldn't `Nothing` be a singleton? maybe it doesn't matter, as long as they compare

Comment (davidchambers) [inline index.js]: I'd like to have just a single value of the Nothing type, but I haven't worked out a solution which meets all of the following requirements:

  - `new S.Nothing()` must return the value;
  - `S.Nothing()` must return the value;
  - applying `R.is(S.Nothing)` to the value must evaluate true; and
  - the `name` property of the value's constructor must be `'Nothing'`.

In fact, I don't think it's possible to meet all these requirements.

Comment (SimonRichardson) [inline index.js]: Don't know if this helps, but we solved some of those issues above for fantasy-land implementions (option, either and etc) with this... https://github.com/puffnfresh/daggy/blob/master/src/daggy.js#L103

Comment (davidchambers) [inline index.js]: Interesting! I don't fully understand that code upon first reading. I'll revisit it this weekend to work out exactly what it does. :)

**Comment by davidchambers** [comment]
I haven't yet merged this as I'm hoping that fantasyland/fantasy-land#90 will be accepted, at which point I'll update the identifiers in the tests to match the updated specification.

**Pull request merged**